Add delete feature.

yichinweng opened the pull request:
contain below features:
1. delete note.
2. if the focus note is deleted, then change focus to the first.
3. if all notes are deleted, generating a default note.

yichinweng committed:
Add delete feature.

contain below features:
1. delete note.
2. if the focus note is deleted, then change focus to the first.
3. if all notes are deleted, generating a default note.

kuoabu commented on src/components/App/App.js:
```diff
@@ -29,10 +29,21 @@ class App extends React.Component {
+      notes = [getDefaultNote()];
+    }
     this.setState({ notes });
+    this.refreshFocus(key);
```
這裡建議放在 setState 第二個參數 callback 裡，setState 不一定會立即和 state 同步

https://facebook.github.io/react/docs/react-component.html#setstate
> setState() does not always immediately update the component. It may batch or defer the update until later. This makes reading this.state right after calling setState() a potential pitfall. Instead, use componentDidUpdate or a setState callback (setState(updater, callback)), either of which are guaranteed to fire after the update has been applied.

yichinweng replied:
感謝您的寶貴提醒，已修正於 #1517bc4。

kuoabu commented on src/components/List/List.js:
```diff
@@ -7,9 +7,18 @@ class List extends React.Component {
   isActive(key) {
     return `note-list-item ${(key === this.props.focus) ? 'active' : ''}`;
   }
+  delete(e, key) {
```
要接 event 的話我喜歡叫 handleDeleteButtonClick XD 
delete 比較像專心處理像呼叫 api 從資料庫移除、或從 array 移除項目的時候

yichinweng replied:
因為之前 content 的 update 都只寫 update，所以我就沿用了XD，
但確實不好。已修正於 #1517bc4。

yichinweng committed:
Code review for #506810d.

[What]
(1) set refreshFocus as callback.
(2) edit delete function name.

[Why]
(1) setState is a request, it may be defer.
(2) delete is not a good name.

[How]
(1) set it as a callback.
(2) change the name to handleDeleteButtonClick.

kuoabu commented on src/components/App/App.js:
```diff
@@ -29,10 +29,20 @@ class App extends React.Component {
+    if (notes.length === 0) {
+      notes = [getDefaultNote()];
+    }
+    this.setState({ notes }, this.refreshFocus(key));
```
這裡要寫，不然 `this.refreshFocus()` 還是會馬上執行 :0!
```
this.setState({ notes }, () => {
  this.refreshFocus(key)
});
```

yichinweng replied:
感謝您的寶貴提醒，已修正於 #b5c6848。

yichinweng committed:
Code review for #1517bc4.

[What]
put callback in arrow function to avoid executing right away.
